node/network/statement-distribution/src/vstaging/grid.rs

```diff
@@ -1518,20 +1503,533 @@ mod tests {
+
+		// Note the manifest as 'sent' to validator 0.
+		tracker.manifest_sent_to(&groups, validator_index, candidate_hash, local_knowledge);
+
+		// Import manifest.
+		//
+		// Should overwrite existing `Full` manifest.
+		let ack = tracker.import_manifest(
+			&session_topology,
+			&groups,
+			candidate_hash,
+			3,
+			ManifestSummary {
+				claimed_parent_hash: Hash::repeat_byte(0),
+				claimed_group_index: group_index,
+				seconded_in_group: bitvec::bitvec![u8, Lsb0; 0, 1, 0],
+				validated_in_group: bitvec::bitvec![u8, Lsb0; 1, 0, 1],
+			},
+			ManifestKind::Acknowledgement,
+			validator_index,
+		);
+		assert_matches!(ack, Ok(false));
+
+		let pending_manifest = tracker.is_manifest_pending_for(validator_index, &candidate_hash);
+		assert_eq!(pending_manifest, None);
+	}
+
+	// Check that pending communication is cleared correctly in `manifest_sent_to`
+	//
+	// Also test a scenario where manifest import returns `Ok(true)`.
+	#[test]
+	fn pending_communication_is_cleared() {
+		let validator_index = ValidatorIndex(0);
+
+		let mut tracker = GridTracker::default();
+		let session_topology = SessionTopologyView {
+			group_views: vec![(
+				GroupIndex(0),
+				GroupSubView {
+					sending: HashSet::new(),
+					receiving: HashSet::from([validator_index]),
+				},
+			)]
+			.into_iter()
+			.collect(),
+		};
+
+		let candidate_hash = CandidateHash(Hash::repeat_byte(42));
+		let group_index = GroupIndex(0);
+		let group_size = 3;
+		let local_knowledge = StatementFilter::new(group_size);
+
+		let groups = dummy_groups(group_size);
+
+		// Add the candidate as backed.
+		tracker.add_backed_candidate(
+			&session_topology,
+			candidate_hash,
+			group_index,
+			group_size,
+			local_knowledge.clone(),
+		);
+
+		// Manifest should not be pending yet.
+		let pending_manifest = tracker.is_manifest_pending_for(validator_index, &candidate_hash);
+		assert_eq!(pending_manifest, None);
+
+		// Import manifest.
+		let ack = tracker.import_manifest(
+			&session_topology,
+			&groups,
+			candidate_hash,
+			3,
+			ManifestSummary {
+				claimed_parent_hash: Hash::repeat_byte(0),
+				claimed_group_index: group_index,
+				seconded_in_group: bitvec::bitvec![u8, Lsb0; 0, 1, 0],
+				validated_in_group: bitvec::bitvec![u8, Lsb0; 1, 0, 1],
+			},
+			ManifestKind::Full,
+			validator_index,
+		);
+		assert_matches!(ack, Ok(true));
+
+		// Acknowledgement manifest should be pending.
+		let pending_manifest = tracker.is_manifest_pending_for(validator_index, &candidate_hash);
+		assert_eq!(pending_manifest, Some(ManifestKind::Acknowledgement));
+
+		// Note the candidate as advertised.
+		tracker.manifest_sent_to(&groups, validator_index, candidate_hash, local_knowledge);
+
+		// Pending manifest should be cleared.
+		let pending_manifest = tracker.is_manifest_pending_for(validator_index, &candidate_hash);
+		assert_eq!(pending_manifest, None);
+	}
+
+	#[test]
+	fn pending_statements_are_updated_after_manifest_exchange() {
```

Comment:
Haven't finished this one yet; what is manifest exchange?

Reply:
A manifest exchange means that both `manifest_sent_to` and `manifest_received_from` have been invoked.

Reply:
In practice, it means that one of three things have happened:
  * They announced, we acknowledged
  * We announced, they acknowledged
  * We announced, they announced (not sure if this can actually happen; it would happen if 2 nodes had each other in their `sending` set _and_ they sent manifests at the same time. the code accounts for this anyway)

Reply:
For the test itself, we don't consider statements as "pending" until the manifest exchange is completed (i.e. both peers recognize the other as having the backed candidate)

Reply:
Thanks! I've tried writing this test, but I can't get it to work. I made sure to call both `manifest_sent_to` and `manifest_received_from` (via `import_manifest`), but pending statements are not updated unless I explicitly call `learned_fresh_statement`. Is this expected, and if so why? I've pushed my code for this test.

Reply:
Ah, I've looked into it more, and it's because we need to have local knowledge of a statement somehow. `learned_fresh_statement` is one way of doing that. I redid the test, should be much more clear now. No bugs as far as I can tell.